### ISSUE TITLE
client.templates fails if  filtered_api = true

### ISSUE THREAD

**Issue by aalba6675** [abenari/rbovirt#110]
ovirt 3.6.5, rbovirt 0.1.1
- User with admin privileges

``` ruby
client = OVIRT::Client.new(
  "adminuser@domain",
  "password",
  "https://ovirt.example.com/api",
  { :datacenter_id =>'5ac1ff95-d7ac-4b42-89d5-241eef81b9a6',
    :ca_no_verify => true
  },)
client.templates ## returns lots of information about templates
```
- Same user 

``` ruby
client = OVIRT::Client.new(
  "adminuser@domain",
  "password",
  "https://ovirt.example.com/api",
  { :datacenter_id =>'5ac1ff95-d7ac-4b42-89d5-241eef81b9a6',
    :ca_no_verify => true, :filtered_api = true
  },)
irb(main):058:0> client.templates
NoMethodError: undefined method `[]' for nil:NilClass
        from /home/vagrant/.gem/ruby/gems/rbovirt-0.1.1/lib/ovirt/template.rb:40:in `parse_xml_attributes!'
        from /home/vagrant/.gem/ruby/gems/rbovirt-0.1.1/lib/ovirt/template.rb:8:in `initialize'
        from /home/vagrant/.gem/ruby/gems/rbovirt-0.1.1/lib/client/template_api.rb:7:in `new'
        from /home/vagrnat/.gem/ruby/gems/rbovirt-0.1.1/lib/client/template_api.rb:7:in `block in templates'
        from /usr/share/gems/gems/nokogiri-1.6.7.2/lib/nokogiri/xml/node_set.rb:187:in `block in each'
        from /usr/share/gems/gems/nokogiri-1.6.7.2/lib/nokogiri/xml/node_set.rb:186:in `upto'
        from /usr/share/gems/gems/nokogiri-1.6.7.2/lib/nokogiri/xml/node_set.rb:186:in `each'
        from /home/vagrant/.gem/ruby/gems/rbovirt-0.1.1/lib/client/template_api.rb:6:in `collect'
        from /home/vagrant/.gem/ruby/gems/rbovirt-0.1.1/lib/client/template_api.rb:6:in `templates'
        from (irb):58
        from /usr/bin/irb:11:in `<main>'
```

**Comment by aalba6675** [comment]
In both cases, ovirt replies but in filtered_api=true rbovirt doesn't seem to parse the data correctly

Working GET ( no filtered_api)

```
.4."....HTTP/1.1.../api/templates..
10.10.10.12.....ovirt..............application/xml....
gzip, deflate....*Basic b3ZpcnR1c2VyQGZ1d2VsbDphYmNkMTIzNA==.....application/xml.....Ruby.....ovirt......."search=+datacenter%3D%22DCENTER%22....ECDHE-RSA-AES128-GCM-SHA256....
..AJP_REMOTE_PORT...36814..AB.c.....OK.....Content-Encoding...gzip.....application/xml.....1886.....Tue, 10 May 2016 09:00:49 GMT.AB.b..^...............6.....@y...`l........qRo...2.I...@.n........|$j/+v_.x<....?y7...6.>..J..~....R..e........laU5.$O.L./.T..~.*...Hy...+K....M.......>...w..P
```

Faulty GET (:filtered_api = true)
```
.4......HTTP/1.1.../api/templates..
10.10.10.12.....ovirt..............application/xml....
gzip, deflate...Filter...true....*Basic b3ZpcnR1c2VyQGZ1d2VsbDphYmNkMTIzNA==.....application/xml.....Ruby.....ovirt........ECDHE-RSA-AES128-GCM-SHA256....
..AJP_REMOTE_PORT...44126..AB.c.....OK.....Content-Encoding...gzip.....application/xml.....2109.....Tue, 10 May 2016 09:01:46 GMT.AB.A..=...............8.....@y..`.1&.du...N.......C.6...9 m...{HH.L.zHTMf.V......|..<...e.Z.tQ&yv;

```

**Comment by aalba6675** [comment]
The difference is due to       `path += search_url(opts) unless unfltered_api` which triggers the error
```ruby
  class Client
    def templates(opts={})
      path = "/templates"
      path += search_url(opts) unless unfltered_api
      result = http_get(path).xpath('/templates/template')
      puts "DEBUG:  #{result.class.name}"
      http_get(path).xpath('/templates/template').collect do |t|
        OVIRT::Template::new(self, t)
      end.compact
    end
```

**Comment by aalba6675** [comment]
Header from ovirt-shell with filtered_api = true
```
.4......HTTP/1.1.."/api/templates;case_sensitive=True..
10.10.10.122.....ovirt.........
....ovirt........*Basic b3ZpcnR1c2VyQGZ1d2VsbDphYmNkMTIzNA==....dPycURL/7.19.5.1 libcurl/7.43.0 NSS/3.22 Basic ECC zlib/1.2.8 libidn/1.32 libssh2/1.6.0 nghttp2/1.7.1....)JSESSIONID=TQ1C8vz2PAEslO7eaS11iPHH.ovirt...Version...3.....application/xml.....application/xml...Filter...True...Prefer...persistent-auth.....0....ECDHE-RSA-AES128-GCM-SHA256...@02091a4a341555c12dba95e3e77494ecb2ac27fdb8c8e1299424cbf901aac5a4....
..AJP_REMOTE_PORT...52522..AB.@.....OK.......application/xml.....Tue, 10 May 2016 07:49:15 GMT.AB.....<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<templates>
    <template href="/api/templates/00000000-0000-00
```

**Comment by aalba6675** [comment]
The :filtered_api = true search results include  the "Blank" template with no <cluster> element, this causes template.rb to fail at 

```ruby   
  def parse_xml_attributes!(xml)
      ## etc etc
      @cluster = Link::new(@client, (xml/'cluster').first[:id], (xml/'cluster').first[:href])
       ## Template "Blank"  xml/'cluster'.first is nil
```
I propose 
```ruby
    def parse_xml_attributes!(xml)
      @description = ((xml/'description').first.text rescue '')
      @comment = ((xml/'comment').first.text rescue '')
      @version = TemplateVersion.new((xml/'version').first)
      @status = ((xml/'status').first.text rescue 'unknown')
      @memory = (xml/'memory').first.text
      @profile = (xml/'type').first.text
      if (xml/'cluster').first
        @cluster = Link::new(@client, (xml/'cluster').first[:id], (xml/'cluster').first[:href])
      end
```

Tested and works with vagrant-ovirt3 1.9.1.

**Comment by lzap** [comment]
> Tested and works with vagrant-ovirt3 1.9.1.

Go ahead and create a PR for this.

-- 
Later,
 Lukas #lzap Zapletal

**Comment by iNecas** [comment]
Resolved by  #111
